### ISSUE TITLE
Slower than `std::collections::HashMap` when keys and values are strings?

### ISSUE THREAD

**Issue by josephrocca** [rust-lang/hashbrown#88]
Hi, I'm very new to rust and just playing around with hash maps and I found that using the code below, the `std::collections::HashMap` is significantly faster than hashbrown. I'm wondering if someone could explain this? Is hashbrown only meant to be faster for certain types of data? Or am I doing something wrong?

```
use hashbrown::HashMap;
//use std::collections::HashMap;
use rand::Rng;
use std::time::Instant;
use numtoa::NumToA;

fn main() {
    let mut m = HashMap::new();
    let mut rng = rand::thread_rng();
    let now = Instant::now();
    let mut i = 0;
    let mut last_time = now.elapsed().as_millis();
    while i < 10000000 {
      let mut key_buf = [0u8; 10];
      let mut value_buf = [0u8; 10];
      let key = rng.gen_range(0, 2u32.pow(31)).numtoa_str(16, &mut key_buf);
      let value = rng.gen_range(0, 2u32.pow(31)).numtoa_str(16, &mut value_buf);
      m.insert(key.to_owned(), value.to_owned());
      i += 1;
      if i%1000000 == 0 {
        let now_time = now.elapsed().as_millis();
        println!("{} {}", i, now_time-last_time);
        last_time = now_time;
      }
    }
    
}
```

Here's my laptop's output for `hashbrown::HashMap`:
```
1000000 943
2000000 1749
3000000 1388
4000000 2739
5000000 1433
6000000 1811
7000000 2548
8000000 3885
9000000 1603
10000000 1852
```

And here's my laptop's output for `std::collections::HashMap`:
```
1000000 709
2000000 738
3000000 513
4000000 1062
5000000 527
6000000 573
7000000 677
8000000 1479
9000000 550
10000000 569
```

**Comment by Amanieu** [comment]
It seems that `FxHash` is generating particularly poor hashes. If you make hashbrown use the libstd default hasher (SipHash) then it generates comparable results:

```rust
let mut m = HashMap::with_hasher(std::collections::hash_map::RandomState::new());
```

**Comment by josephrocca** [comment]
Ah okay, I just wasn't sure whether there was something strange going on, or whether this is expected/not-surprising. I figured that since SiphHash is DOS-resistant, and since I don't need that security for my use case, then I could go with hashbrown and get a bit of a speedup. I'll just stick with the built-in HashMap unless I'm dealing with integers. Thanks for your help!

**Comment by Amanieu** [comment]
I'd like to keep this open, since we really need to fix the hasher.

**Comment by josephrocca** [comment]
Ah, great! I'm too new to to rust to help out myself at this point, but I'm looking forward to testing out the improvements!

**Comment by xacrimon** [comment]
Fxhash is know to generate poor hashes often. I'd consider switching to Seahash. It's what I did I'm my crate ccl.

**Comment by Amanieu** [comment]
Fixed by #97
